src/intersections/boxes.jl

```diff
@@ -34,3 +34,51 @@ function intersecttype(f::Function, b1::Box{Dim,T}, b2::Box{Dim,T}) where {Dim,T
     return FaceTouchingBoxes(Box(u, v)) |> f
   end
 end
+
+"""
+    intersecttype(function, ray, box)
+
+Calculate the intersection type of a `ray`` and a `box`` and apply function `f` to it.
+
+## References
+
+* Williams A, Barrus S, Morley R K, et al., 2005. [An efficient and robust ray-box
+  intersection algorithm]
+  (https://dl.acm.org/doi/abs/10.1145/1198555.1198748)
+"""
+function intersecttype(f::Function, r::Ray{Dim,T}, b::Box{Dim,T}) where {Dim,T}
+  invdir = one(T) ./ direction(r)
+  lo, up = coordinates.(extrema(b))
+  orig = coordinates(origin(r))
+
+  tmin = zero(T)
+  tmax = typemax(T)
+
+  for i in 1:Dim
+    if invdir[i] ≥ zero(T)
+      imin = (lo[i] - orig[i]) * invdir[i]
+      imax = (up[i] - orig[i]) * invdir[i]
+    else
+      imin = (up[i] - orig[i]) * invdir[i]
+      imax = (lo[i] - orig[i]) * invdir[i]
+    end
```

Comment:
Notice that now that the code is clean we can see the logic here as "swap imin and imax if the sign of invdir is negative". I would erase this code repetition and would write it as:

```julia
for i in 1:Dim
  imin = (lo[i] - orig[i]) * invdir[i]
  imax = (up[i] - orig[i]) * invdir[i]
  
  # swap if necessary
  invdir[i] < zero(T) && ((imin, imax) = (imax, imin))

  # continue the code...
```